kernel/work_q.c

```diff
@@ -116,18 +114,18 @@ int k_delayed_work_cancel(struct k_delayed_work *work)
 {
 	int key = irq_lock();
 
-	if (k_work_pending(&work->work)) {
-		irq_unlock(key);
-		return -EINPROGRESS;
-	}
-
 	if (!work->work_q) {
 		irq_unlock(key);
 		return -EINVAL;
 	}
 
-	/* Abort timeout, if it has expired this will do nothing */
-	_abort_timeout(&work->timeout);
+	if (k_work_pending(&work->work)) {
+		/* Remove from the queue if already submitted */
+		k_queue_find_and_remove(&work->work_q->queue, work);
```

Comment:
Both using k_sem or k_mutex seems to have not a problem with ISR, though they both rely on _wait_q_t/_pend_current_thread which does list lookup with irq_lock, so there doesn't seem to be a way around it. Also _abort_timeout appear to have a similar design, though it uses sys_dlist_t which has a fixed bound for sys_dlist_remove since it has the pointer to the previous item. IMO the only way to fix this aspect is to convert k_work and k_work_q to use sys_dlist_t, though this will increase the memory of every single work item, so perhaps we need to measure this first with sys_slist_ and if not acceptable then do convert to sys_dlist_t.